controller/controller.go

```diff
@@ -232,6 +232,52 @@ func (c *notificationController) processApp(app *unstructured.Unstructured, logE
 	return nil
 }
 
+// Checks if the application SyncStatus has been refreshed by Argo CD after an operation has completed
+func (c *notificationController) isAppSyncStatusRefreshed(app *unstructured.Unstructured, logEntry *log.Entry) bool {
+	_, ok, err := unstructured.NestedMap(app.Object, "status", "operationState")
+	if !ok || err != nil {
+		logEntry.Debug("No OperationState found, SyncStatus is assumed to be up-to-date")
+		return true
+	}
+
+	phase, ok, err := unstructured.NestedString(app.Object, "status", "operationState", "phase")
+	if !ok || err != nil {
+		logEntry.Debug("No OperationPhase found, SyncStatus is assumed to be up-to-date")
+		return true
+	}
+	switch phase {
+	case "Failed", "Error", "Succeeded":
+		finishedAtRaw, ok, err := unstructured.NestedString(app.Object, "status", "operationState", "finishedAt")
+		if !ok || err != nil {
+			logEntry.Debugf("No FinishedAt found for completed phase '%s', SyncStatus is assumed to be out-of-date", phase)
```

Comment:
I think notifications should be skipped only for a short period of time after sync operation completion and before the next reconciliation. So if sync is in progress and finishedAt is missing we still need to send notifications.

Reply:
Could you explain in more depth what you mean please?

As far as I understand the argo-cd code, the phase and finishedAt are set [at the same time](https://github.com/argoproj/argo-cd/blob/master/controller/appcontroller.go#L733) for [completed phases](https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L484).

Of course `unstructured.NestedString` does not know this and we have to check the `ok` and `err`. Maybe the debug msg should be different, but it is a (in)famous 'this should never happen' msg I guess.

If the sync is in progress the phase should be `Running`, and we should not hit the finishedAt check at all. For the phases running and terminating we just assume the SyncStatus is up to date and send out notifications right away. Only for completed phases (failed, error and succeeded) we delay sending the notification till argo-cd has updated the SyncStatus by checking the observedAt and reconciledAt timestamps.

Reply:
Oh, you are right, I've missed that `finishedAt` is verified after checking the operation phase. 👍